src/internal/clusterstate/v2.8.0/defaults.go

```diff
@@ -123,3 +128,116 @@ func envMap(environ []string) (map[string]string, error) {
 	}
 	return env, nil
 }
+
+func versionKey(p *pps.Pipeline, version uint64) string {
+	// zero pad in case we want to sort
+	return fmt.Sprintf("%s@%08d", p, version)
+}
+
+var pipelinesVersionIndex = &index{
+	Name: "version",
+	Extract: func(val proto.Message) string {
+		info := val.(*pps.PipelineInfo)
+		return versionKey(info.Pipeline, info.Version)
+	},
+}
+
+func pipelinesNameKey(p *pps.Pipeline) string {
+	return p.String()
+}
+
+var pipelinesNameIndex = &index{
+	Name: "name",
+	Extract: func(val proto.Message) string {
+		return pipelinesNameKey(val.(*pps.PipelineInfo).Pipeline)
+	},
+}
+
+var pipelinesIndexes = []*index{
+	pipelinesVersionIndex,
+	pipelinesNameIndex,
+}
+
+func pipelineCommitKey(commit *pfs.Commit) (string, error) {
+	if commit.Repo.Type != pfs.SpecRepoType {
+		return "", errors.Errorf("commit %s is not from a spec repo", commit)
+	}
+	if projectName := commit.Repo.Project.GetName(); projectName != "" {
+		return fmt.Sprintf("%s/%s@%s", projectName, commit.Repo.Name, commit.Id), nil
+	}
+	return fmt.Sprintf("%s@%s", commit.Repo.Name, commit.Id), nil
+}
+
+func withKeyGen(gen func(interface{}) (string, error)) colOption {
+	return func(c *postgresCollection) {
+		c.keyGen = gen
+	}
+}
+
+func withKeyCheck(check func(string) error) colOption {
+	return func(c *postgresCollection) {
+		c.keyCheck = check
+	}
+}
+
+func parsePipelineKey(key string) (projectName, pipelineName, id string, err error) {
+	parts := strings.Split(key, "@")
+	if len(parts) != 2 || !uuid.IsUUIDWithoutDashes(parts[1]) {
+		return "", "", "", errors.Errorf("key %s is not of form [<project>/]<pipeline>@<id>", key)
+	}
+	id = parts[1]
+	parts = strings.Split(parts[0], "/")
+	if len(parts) == 0 {
+		return "", "", "", errors.Errorf("key %s is not of form [<project>/]<pipeline>@<id>")
+	}
+	pipelineName = parts[len(parts)-1]
+	if len(parts) == 1 {
+		return
+	}
+	projectName = strings.Join(parts[0:len(parts)-1], "/")
+	return
+}
+
+func synthesizeSpec(pi *pps.PipelineInfo) (*pps.PipelineInfo, error) {
+	// create an initial user and effective spec equal to what would have been previously used
+	spec := ppsutil.PipelineReqFromInfo(pi)
+	js, err := protojson.Marshal(spec)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not marshal CreatePipelineRequest as JSON")
+	}
+	if pi.EffectiveSpecJson == "" {
+		pi.EffectiveSpecJson = string(js)
+	}
+	if pi.UserSpecJson == "" {
+		pi.UserSpecJson = string(js)
+	}
+	return pi, nil
+}
+
+func synthesizeSpecs(ctx context.Context, env migrations.Env) error {
+	var oldPipeline = new(pps.PipelineInfo)
+	if err := migratePostgreSQLCollection(ctx, env.Tx, "pipelines", pipelinesIndexes, oldPipeline, func(oldKey string) (newKey string, newVal proto.Message, err error) {
+		if oldPipeline, err = synthesizeSpec(oldPipeline); err != nil {
```

Comment:
```suggestion
		if newPipeline, err = synthesizeSpec(oldPipeline); err != nil {
```

Reply:
I’ve updated this to just update in place.